src/plots/cartesian/index.js

```diff
@@ -224,43 +224,50 @@ exports.clean = function(newFullData, newFullLayout, oldFullData, oldFullLayout)
     var oldModules = oldFullLayout._modules || [],
         newModules = newFullLayout._modules || [];
 
-    var hadScatter, hasScatter, hadGl, hasGl, i, oldPlots, ids, subplotInfo, moduleName;
+    var hadScatter, hasScatter, hadOHLC, hasOHLC, hadGl, hasGl, i, oldPlots, ids, subplotInfo, moduleName;
```

Comment:
This had/has logic is also getting ridiculous. 

What we need to do is find all the _gone_ modules (i.e. modules with `had??=true` and `has??=false`), loop over them, and then do what we need to do the cleanup the DOM. 

Better yet -- as more and more trace type will follow scatter and ohlc in not expecting a `selectAll('g.trace').remove()` before each `_module.plot()` -- we could add a trace module method (maybe called `cleanDOM`) to each trace modules that need to remove things for the DOM (or destroy WebGL objects). 

To note, geo subplots (and maybe others?) solve this problem in a different way by using `Plots.generalUpdatePerTraceModule` which calls `_module.plot` of visible **and**  _gone_ modules and expects the trace module plot methods to clean up the DOM accordingly (e.g. like [here](https://github.com/plotly/plotly.js/blob/18e19a86a3e683c1d11f6f4a04eb7f89d1b6108f/src/traces/choropleth/plot.js#L34) for choropleth traces). At present though, I think adding `cleanDOM` to each trace module would be best, as it would allow us to handle the `scattergl` and range-slider cases more easily.

N.B. the splom PR (ref: https://github.com/plotly/plotly.js/pull/2505) also patches this code here, so maybe addressing this comment could be done in a different PR

Reply:
Definitely a separate PR - there will be a bit to discuss, between this one and https://github.com/plotly/plotly.js/pull/2561#discussion_r181852562 - created #2563, OK to defer both of these for now?

Reply:
> OK to defer both of these for now?

Absolutely :ok_hand: